checker/src/org/checkerframework/checker/index/upperbound/UpperBoundAnnotatedTypeFactory.java

```diff
@@ -493,12 +495,36 @@ public Void visitBinary(BinaryTree tree, AnnotatedTypeMirror type) {
                 case DIVIDE:
                     addAnnotationForDivide(left, right, type);
                     break;
+                case AND:
+                    addAnnotationForAnd(left, right, type);
+                    break;
                 default:
                     break;
             }
             return super.visitBinary(tree, type);
         }
 
+        private void addAnnotationForAnd(
+                ExpressionTree left, ExpressionTree right, AnnotatedTypeMirror type) {
+            AnnotatedTypeMirror leftType = getAnnotatedType(left);
+            AnnotatedTypeMirror leftLBType =
+                    getLowerBoundAnnotatedTypeFactory().getAnnotatedType(left);
+            if (leftLBType.hasAnnotation(NonNegative.class)
+                    || leftLBType.hasAnnotation(Positive.class)) {
+                type.addAnnotation(leftType.getAnnotationInHierarchy(UNKNOWN));
+            }
+
+            AnnotatedTypeMirror rightType = getAnnotatedType(right);
+            AnnotatedTypeMirror rightLBType =
+                    getLowerBoundAnnotatedTypeFactory().getAnnotatedType(right);
+            if (rightLBType.hasAnnotation(NonNegative.class)
+                    || rightLBType.hasAnnotation(Positive.class)) {
+                type.addAnnotation(rightType.getAnnotationInHierarchy(UNKNOWN));
```

Comment:
This code doesn't account for the case when the left and right expressions are indexes for two different arrays:
````
int[] a;
int[] b;
void test(@IndexFor("a") int i, @IndexFor("b") int j) {
  a[i & j] = 1;
  b[i & j] = 2;
}
````